plugins/GSdx/GSTextureCache.cpp

```diff
@@ -2043,3 +2033,119 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
 
 	delete s;
 }
+
+// Query the PaletteMap for a valid Palette, then assign both palette texture pointer and clut copy pointer to the Source object
+void GSTextureCache::SetPalette(Source* s, const GSRenderer* renderer, uint16 pal)
+{
+	Palette* p = m_palette_map.GetPalette(m_renderer, pal);
+	s->m_palette = p->GetPaletteGSTexture();
+	s->m_clut = p->GetClut();
+}
+
+// GSTextureCache::Palette
+
+// Default constructor
+GSTextureCache::Palette::Palette() {
+	m_clut = nullptr;
+	m_tex_palette = nullptr;
+}
+
+// Copy constructor
+GSTextureCache::Palette::Palette(const Palette& palette) {
+	m_clut = palette.m_clut;
+	m_tex_palette = palette.m_tex_palette;
+};
+
+// Creates a new palette texture with current clut content, keeping a reference to its copy
+GSTextureCache::Palette::Palette(const GSRenderer* renderer, uint16 pal) {
+	uint16 palette_size = pal * sizeof(uint32);
+	m_clut = (uint32*)_aligned_malloc(palette_size, 64);
+	memcpy(m_clut, (const uint32*)renderer->m_mem.m_clut, palette_size);
+	m_tex_palette = renderer->m_dev->CreateTexture(256, 1);
+	m_tex_palette->Update(GSVector4i(0, 0, pal, 1), m_clut, palette_size);
+}
+
+uint32* GSTextureCache::Palette::GetClut() {
+	return m_clut;
+}
+
+GSTexture* GSTextureCache::Palette::GetPaletteGSTexture() {
+	return m_tex_palette;
+}
+
+// GSTextureCache::PaletteMap
+
+// Default constructor, reverses space in the multimaps
+GSTextureCache::PaletteMap::PaletteMap() {
+	for (auto& multimap : m_multimaps) {
+		multimap.reserve(MAX_SIZE);
+	}
+}
+
+// Hashes the content of the clut
+size_t GSTextureCache::PaletteMap::HashClut(uint16 pal, const uint32* clut) {
+	size_t clut_hash = 3831179159;
```

Comment:
Add some comments to explain where you did get those hash value.

Reply:
Big prime number out of the magic hat 🎩 
Jokes aside, I tried to build a simple hash function and this seemed to work well.
It can be improved by hashing 16 indices at a time, using different prime numbers (they have magic properties when hashing) for each index.
The hash function is critical and open to improvements.
I didn't find any SSE method which is "easy" to implement for hashing.
The use case is very specialized though, so maybe hashing 16 values (1 iteration for pal == 16, 16 iterations for pal == 256) at a time with different combinations of operations/prime numbers may give good results.

Reply:
Comment will be nice. This way, people can test various hash function to see the impact.

Reply:
Also I prefer if void and similar functions being in a new line like this
```
void something
{
}
```
instead of `void something (`